File: src/shared/types/theme.ts
/**
 * Theme domain types shared between frontend and backend.
 * Mirrors backend/schemas/theme.py.
 */

export interface ThemePreset {
  id: string; // 'light' | 'dark' | 'ocean' | 'forest' | 'sunset' | user-*
  name: string; // i18n key
  description: string; // i18n key
  cover?: string; // URL or relative path
  css?: string; // optional raw CSS for user-saved themes
}

export interface ThemeCssPayload {
  css: string; // raw CSS string
  vars: Record<string, string>; // parsed 16-var whitelist
}

export interface ActiveTheme {
  presetId: string;
  customCss?: string;
}

export interface ThemeValidationResult {
  valid: boolean;
  errors?: string[];
}

/**
 * The 16 CSS variables that theme CSS is allowed to override.
 * Mirrors backend ALLOWED_CSS_VARS in backend/api/theme_router.py.
 * Adding/removing vars requires updating BOTH sides + adding test cases.
 */
export const ALLOWED_CSS_VARS = [
  '--color-bg',
  '--color-bg-secondary',
  '--color-bg-tertiary',
  '--color-fg',
  '--color-fg-secondary',
  '--color-fg-muted',
  '--color-border',
  '--color-border-strong',
  '--color-accent',
  '--color-accent-hover',
  '--color-success',
  '--color-warning',
  '--color-error',
  '--color-info',
  '--color-link',
  '--color-link-hover',
] as const;

export type AllowedCssVar = (typeof ALLOWED_CSS_VARS)[number];

File: backend/schemas/theme.py
"""Theme domain models (pydantic 2.x compatible)."""

from pydantic import BaseModel, ConfigDict, Field


class ThemePreset(BaseModel):
    """A named theme preset - builtin or user-created."""

    model_config = ConfigDict(extra="forbid")

    id: str = Field(..., pattern=r"^[a-z0-9-]{1,32}$")
    name: str  # i18n key
    description: str  # i18n key
    cover: str | None = None
    css: str | None = None


class ThemeCssPayload(BaseModel):
    """Raw CSS string + parsed whitelisted variables."""

    model_config = ConfigDict(extra="forbid")

    css: str
    vars: dict[str, str] = {}  # 16-var whitelist, parsed separately by the API layer


class ActiveTheme(BaseModel):
    """Currently active theme: a preset + optional custom CSS overlay."""

    model_config = ConfigDict(extra="forbid")

    presetId: str
    customCss: str | None = None
